### PR TITLE
Add hardware-watchdog abstraction.

### PR DESCRIPTION
**Pull request by floitsch**

### CODE REVIEW

File: src/provider.toit [hunk 1]
```diff
@@ -2,24 +2,62 @@
...
 // in the LICENSE file.
 
-import esp32
 import log
 import monitor
 import system.services show ServiceProvider ServiceResource ServiceHandler
 import .api.service
+import .esp32 show HardwareWatchdogEsp32
+
+interface HardwareWatchdog:
```
Comment (kasperl) [inline src/provider.toit:11]: I'm not sure hardware is a great name here given that this can be mocked, etc.

**Review by kasperl — COMMENTED** [review]
Looks good to me.

I think I prefer System over Hardware, but it is nice to have an abstraction for this so it can be tested better.

File: src/provider.toit (hunk 1 above)
Comment (floitsch) [reply src/provider.toit]: Renamed back to "system" (`SystemWatchdog`)

File: src/provider.toit [hunk 2]
```diff
@@ -2,24 +2,62 @@
...
 import system.services show ServiceProvider ServiceResource ServiceHandler
 import .api.service
+import .esp32 show SystemWatchdogEsp32
+
+interface SystemWatchdog:
+  /**
+  Starts the hardware watchdog timer.
+
+  The watchdog will reset the system if it isn't fed within $ms milliseconds.
+  The hardware watchdog may have a coarser granularity than the given $ms.
```
Comment (kasperl) [inline src/provider.toit:16]:
```suggestion
  The system watchdog may have a coarser granularity than the given $ms.
```

File: src/provider.toit [hunk 3]
```diff
@@ -2,24 +2,62 @@
...
+  */
+  start --ms/int
+
+  /**
+  Feeds the hardware watchdog.
```
Comment (kasperl) [inline src/provider.toit:21]:
```suggestion
  Feeds the system watchdog.
```

File: src/provider.toit [hunk 4]
```diff
@@ -2,24 +2,62 @@
...
+
+  # Aliases
+  - `kick`
+  - `reset`
+  */
+  feed -> none
+
+  /**
+  Stops the hardware watchdog.
```
Comment (kasperl) [inline src/provider.toit:30]:
```suggestion
  Stops the system watchdog.
```

File: src/provider.toit [hunk 5]
```diff
@@ -2,24 +2,62 @@
...
-import esp32
 import log
 import monitor
 import system.services show ServiceProvider ServiceResource ServiceHandler
 import .api.service
+import .esp32 show SystemWatchdogEsp32
+
+interface SystemWatchdog:
+  /**
+  Starts the hardware watchdog timer.
```
Comment (kasperl) [inline src/provider.toit:13]:
```suggestion
  Starts the system watchdog timer.
```

File: src/provider.toit [hunk 6]
```diff
@@ -66,24 +104,27 @@ class WatchdogServiceProvider extends ServiceProvider
     if system-watchdog-task_: return
 
     mutex_.do:
-      logger_.debug "starting system watchdog"
-      esp32.watchdog-init --ms=GRANULARITY-MS
+      logger_.debug "starting hardware watchdog"
```
Comment (kasperl) [inline src/provider.toit:107]:
```suggestion
      logger_.debug "starting system watchdog"
```

File: src/provider.toit [hunk 7]
```diff
@@ -66,24 +104,27 @@ class WatchdogServiceProvider extends ServiceProvider
...
+              // Feed the hardware watchdog one last time then request to reboot.
+              // This allows the system to clean up before rebooting.
+              mutex_.do: system-watchdog_.feed
+              system-watchdog_.reboot
             else:
-              // Feed the system watchdog.
-              logger_.debug "feeding system watchdog"
-              mutex_.do: esp32.watchdog-reset
-              sleep --ms=(GRANULARITY-MS / 2)
+              // Feed the hardware watchdog.
```
Comment (kasperl) [inline src/provider.toit:124]:
```suggestion
              // Feed the system watchdog.
```

File: src/provider.toit [hunk 8]
```diff
@@ -93,10 +134,10 @@ class WatchdogServiceProvider extends ServiceProvider
     needs-watching := dogs_.any: | _ dog/Watchdog | not dog.is-stopped
     if needs-watching: return
 
-    // Shutdown the system watchdog.
+    // Shutdown the hardware watchdog.
```
Comment (kasperl) [inline src/provider.toit:137]:
```suggestion
    // Shutdown the system watchdog.
```

File: src/provider.toit [hunk 9]
```diff
@@ -93,10 +134,10 @@ class WatchdogServiceProvider extends ServiceProvider
...
     mutex_.do:
-      logger_.info "stopping system watchdog"
-      esp32.watchdog-deinit
+      logger_.info "stopping hardware watchdog"
```
Comment (kasperl) [inline src/provider.toit:139]:
```suggestion
      logger_.info "stopping system watchdog"
```

**Review by kasperl — APPROVED** [review]
Looks even better!

Not sure if I missed more of the 'hardware' references...

**Comment by floitsch** [comment]
Thanks. one reference to "hardware" was left.

**Pull request merged**